File: src/main/scala/com/scala/test/KafkaStream.scala
package com.scala.test

import java.util.Properties

import com.scala.test.model.{ProduitBrut, ProduitEnrichi, Referentiel}
import com.scala.test.serializer.{JsonDeserializer, JsonSerializer}
import org.apache.kafka.common.serialization.{Serde, Serdes}
import org.apache.kafka.streams.kstream._
import org.apache.kafka.streams.{KafkaStreams, KeyValue, StreamsConfig}
import org.slf4j.LoggerFactory

class KafkaStream {
}

object KafkaStream {

  val REFERENTIEL_STORE = "referentiel_store"
  val BOOTSTRAP_SERVER = "localhost:9092"
  val TOPIC_ACHATS = "achats"
  val TOPIC_REFERENTIEL = "referentiel"
  val TOPIC_ACHATS_BY_PRODUCT_ID: String = "achats-by-product-id"
  val TOPIC_ACHATS_ENRICHIS = "achats-enrichis"
  val APPLICATION_ID = "enrichissement-achats"

  val logger = LoggerFactory.getLogger(classOf[KafkaStream])

  def main(args: Array[String]) {

    val streamsConfiguration = new Properties()
    streamsConfiguration.put(StreamsConfig.APPLICATION_ID_CONFIG, APPLICATION_ID)
    streamsConfiguration.put(StreamsConfig.BOOTSTRAP_SERVERS_CONFIG, BOOTSTRAP_SERVER)

    val builder = new KStreamBuilder()

    val achatBrutSerde: Serde[ProduitBrut] = Serdes.serdeFrom(new JsonSerializer(), new JsonDeserializer[ProduitBrut])
    val achatEnrichiSerde: Serde[ProduitEnrichi] = Serdes.serdeFrom(new JsonSerializer(), new JsonDeserializer[ProduitEnrichi])
    val referentielSerde: Serde[Referentiel] = Serdes.serdeFrom(new JsonSerializer(), new JsonDeserializer[Referentiel])

    val achats: KStream[String, ProduitBrut] = builder.stream(Serdes.String(), achatBrutSerde, TOPIC_ACHATS)

    val referentiel: KTable[String, Referentiel] = builder.table(Serdes.String(), referentielSerde, TOPIC_REFERENTIEL, REFERENTIEL_STORE)

    val enriched = achats
      .filter((k, v) => v != null)
      .map((k, v) => new KeyValue(v.id.toString, v))
      .through(Serdes.String(), achatBrutSerde, TOPIC_ACHATS_BY_PRODUCT_ID)
      .leftJoin(referentiel, (achat, ref: Referentiel) => {
        if (ref == null) {
          val produitEnrichi = ProduitEnrichi(achat.id, "REF INCONNUE", achat.price)
          logger.info(produitEnrichi.toString)
          produitEnrichi
        } else {
          val produitEnrichi = ProduitEnrichi(achat.id, ref.name, achat.price)
          logger.info(produitEnrichi.toString)
          produitEnrichi
        }
      })

    //    enriched.to(Serdes.String(), achatEnrichiSerde, TOPIC_ACHATS_ENRICHIS)

    val streams: KafkaStreams = new KafkaStreams(builder, streamsConfiguration)
    streams.start()

    //    new KafkaRestService(streams).start
  }
}

File: src/main/java/com/java/test/KafkaStream.java
package com.java.test;

import com.java.test.model.ProduitBrut;
import com.java.test.model.ProduitEnrichi;
import com.java.test.model.Referentiel;
import com.java.test.rest.KafkaRestService;
import com.java.test.serializer.JsonDeserializer;
import com.java.test.serializer.JsonSerializer;
import org.apache.kafka.common.serialization.Serde;
import org.apache.kafka.common.serialization.Serdes;
import org.apache.kafka.streams.KafkaStreams;
import org.apache.kafka.streams.KeyValue;
import org.apache.kafka.streams.StreamsConfig;
import org.apache.kafka.streams.kstream.KStream;
import org.apache.kafka.streams.kstream.KStreamBuilder;
import org.apache.kafka.streams.kstream.KTable;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.Properties;

public class KafkaStream {

    public static final String REFERENTIEL_STORE = "referentiel_store";
    private static final String BOOTSTRAP_SERVER = "localhost:9092";
    private static final String TOPIC_ACHATS = "achats";
    private static final String TOPIC_REFERENTIEL = "referentiel";
    private static final String TOPIC_ACHATS_BY_PRODUCT_ID = "achats-by-product-id";
    private static final String TOPIC_ACHATS_ENRICHIS = "achats-enrichis";
    private static final String APPLICATION_ID = "enrichissement-achats";

    private final static Logger LOGGER = LoggerFactory.getLogger(KafkaStream.class);

    public static void main(String[] args) throws Exception {

        Properties streamsConfiguration = new Properties();
        streamsConfiguration.put(StreamsConfig.APPLICATION_ID_CONFIG, APPLICATION_ID);
        streamsConfiguration.put(StreamsConfig.BOOTSTRAP_SERVERS_CONFIG, BOOTSTRAP_SERVER);

        KStreamBuilder builder = new KStreamBuilder();

        Serde<ProduitBrut> achatBrutSerde = Serdes.serdeFrom(new JsonSerializer<>(), new JsonDeserializer<>(ProduitBrut.class));
        Serde<ProduitEnrichi> achatEnrichiSerde = Serdes.serdeFrom(new JsonSerializer<>(), new JsonDeserializer<>(ProduitEnrichi.class));
        Serde<Referentiel> referentielSerde = Serdes.serdeFrom(new JsonSerializer<>(), new JsonDeserializer<>(Referentiel.class));

        KStream<String, ProduitBrut> achats = builder.stream(Serdes.String(), achatBrutSerde, TOPIC_ACHATS);

        KTable<String, Referentiel> referentiel = builder.table(Serdes.String(), referentielSerde, TOPIC_REFERENTIEL, REFERENTIEL_STORE);

        KStream<String, ProduitEnrichi> enriched = achats
                .filter((k, v) -> v != null)
                .map((k, v) -> new KeyValue<>(v.getId().toString(), v))
                .through(Serdes.String(), achatBrutSerde, TOPIC_ACHATS_BY_PRODUCT_ID)
                .leftJoin(referentiel, (achat, ref) -> {
                    if (ref == null) {
                        ProduitEnrichi produitEnrichi = new ProduitEnrichi(achat.getId(), "REF INCONNUE", achat.getPrice());
                        LOGGER.info(produitEnrichi.toString());
                        return produitEnrichi;
                    } else {
                        ProduitEnrichi produitEnrichi = new ProduitEnrichi(achat.getId(), ref.getName(), achat.getPrice());
                        LOGGER.info(produitEnrichi.toString());
                        return produitEnrichi;
                    }
                });

        enriched.to(Serdes.String(), achatEnrichiSerde, TOPIC_ACHATS_ENRICHIS);

        KafkaStreams streams = new KafkaStreams(builder, streamsConfiguration);
        streams.start();

        new KafkaRestService(streams).start();
    }
}
